pkg/cmd/experimental/sync-groups/groupsyncer.go

```diff
@@ -0,0 +1,284 @@
+			}
+		case GroupSyncSourceOpenShift:
+			groups, err := getOpenShiftGroups(config.WhitelistContents, config.GroupClient)
+			if err != nil {
+				return nil, err
+			}
+			syncer.groupLister = &LocalGroupLister{
+				Groups: groups,
+			}
+		default:
+			return nil, fmt.Errorf("group sync source %v not recognized", config.Source)
+		}
+	case GroupSyncScopeAll:
+		switch config.Souce {
+		case GroupSyncSourceOpenShift:
+			syncer.GroupLister = &AllLocalGroupLister{
+				client:       config.GroupClient,
+				serverConfig: config.ServerConfig,
+			}
+		}
+	default:
+		return nil, fmt.Errorf("group sync scope %v not recognized", config.Scope)
+	}
+
+	// Build and populate the schema-specific portions of the GroupSyncer
+	switch config.Type {
+	case LDAPSyncTypeFirstClassGroups:
+		return newFirstClassGroupsSyncer(syncer)
+	case LDAPSyncTypeFirstClassUserss:
+		return newFirstClassUsersSyncer(syncer)
+	case LDAPSyncTypeFirstClassGroupsAndUsers:
+		return newFirstClassGroupsAndUsersSyncer(syncer)
+	default:
+		return nil, fmt.Errorf("%s is an invalid type of LDAP group syncer", config.Type)
+	}
+}
+
+func getOpenShiftGroups(names []string, client client.GroupsInterface) ([]*ouserapi.Group, err) {
+	var groups []*ouserapi.Group
+	for _, name := range names {
+		group, err := client.Get(name)
+		if err != nil {
+			return nil, err
+		}
+		groups = append(groups, group)
+	}
+	return groups, nil
+}
+
+// newFirstClassGroupsSyncer fleshes out a skeleton LDAPGroupSyncer with the moving parts
+// specific to the first-class LDAP group entry schema
+func newFirstClassGroupsSyncer(syncer LDAPGroupSyncer) (*LDAPGroupSyncer, error) {
+	var nameMapper LDAPGroupNameMapper
+	// Attempt to populate the groupNameMapper if a user-defined mapping is given
+	if config.UserDefinedGroupNameMapping {
+		nameMapper = &UserDefinedLDAPGroupNameMapper{
+			nameMapping: syncer.config.LDAPGroupUIDToOpenShiftNameMapping,
+		}
+	}
+
+	// build the dataExtractor
+	var dataExtractor LDAPGroupDataExtractor
+	switch syncer.config.Scope {
+	case GroupSyncScopeAll:
+		extractor := schema1.GroupListerDataExtractor{
+			ServerConfig:               config.ServerConfig,
+			AllGroupQueryConfig:        config.FirstClassGroupOptions.SpecificGroupQueryConfig.CommonConfig,
+			SpecificUserQueryConfig:    config.FirstClassGroupOptions.SpecificUserQueryConfig,
+			GroupFieldAttributeMapping: config.FirstClassGroupOptions.GroupFieldAttributeMapping,
+			UserToIdentityMapping:      config.LDAPUserToIdentityMapping,
+			cachedUsers:                make(map[string]*ldap.Entry),
+			cachedGroups:               make(map[string]*ldap.Entry),
+		}
+
+		// create a groupNameMapper if the user did not define a mapping
+		if !config.UserDefinedGroupNameMapping {
+			nameMapper = &EntryAttributeLDAPGroupNameMapper{
+				mapping:     config.FirstClassGroupOptions.GroupFieldAttributeMapping,
+				groupGetter: extractor,
+			}
+		}
+		extractor.groupNameMapper = nameMapper
```

Comment:
When we wanted to break out a `GroupNameMapper` as a discrete object and allow us to substitute whatever for it, we run into this cyclical dependency where you can't create the `DataExtractor` without populating the `GroupNameMapper` but the `GroupNameMapper` needs  to have a pointer to a `groupGetter`, which is the `DataExtractor` itself.